doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -438,25 +497,50 @@ make it impossible to implement this generic fallback behavior for
-and performance degradation. 2. Use another sentinel value of some sort
-to indicate that a class implementing part of the higher level array API
-is coercible as a fallback, e.g., a return value of
-``np.NotImplementedButCoercible`` from ``__array_function__``.
-
-If we take this second approach, we would need to define additional
-rules for how coercible array arguments are coerced, e.g., - Would we
-try for ``__array_function__`` overloads again after coercing coercible
-arguments? - If so, would we coerce coercible arguments one-at-a-time,
-or all-at-once?
-
-These are slightly tricky design questions, so for now we propose to
-defer this issue. We can always implement
+``__array_function__`` in either of two possible ways:
+
+1. Change the meaning of all arguments returning ``NotImplemented`` to indicate
+   that all arguments should be coerced to NumPy arrays instead. However, many
+   array libraries (e.g., scipy.sparse) really don't want implicit conversions
+   to NumPy arrays, and often avoid implementing ``__array__`` for exactly this
+   reason. Implicit conversions can result in silent bugs and performance
+   degradation.
+2. Use another sentinel value of some sort, e.g.,
+   ``np.NotImplementedButCoercible``, to indicate that a class implementing part
+   of NumPy's higher level array API is coercible as a fallback. This is a more
+   appealing option.
+
+If we take this second approach, we would need to define additional rules for
+how coercible array arguments are coerced. For sanity, a return value of
+``np.NotImplementedButCoercible`` from ``__array_function__`` should be handled
+equivalently to the corresponding argument not defining an
+``__array_function__`` method at all. If all arguments to a NumPy function
+either have no ``__array_function__`` method defined or return
+``NotImplementedButCoercible`` from ``__array_function__``, we would fall back
+to NumPy's current behavior: all arguments would be coerced to NumPy arrays.
+
+This approach also suggests an intriguing possibility: default implementations
+of ``__array_function__`` and ``__array_ufunc__`` on ``numpy.ndarray`` could
+change to simply return ``NotImplementedButCoercible``, i.e.,
+
+.. code:: python
+
+    class ndarray:
+        def __array_ufunc__(self, *args, **kwargs):
+            return NotImplementedButCoercible
+        def __array_function__(self, *args, **kwargs):
+            return NotImplementedButCoercible
+
+This logic is not only much simpler than the existing implementation of
+``numpy.ndarray.__array_function__``, but it also means we could remove all
+special cases for ``numpy.ndarray`` in the NumPy's dispatch logic. It would
+break `some existing implementations <https://mail.python.org/pipermail/numpy-discussion/2018-June/078175.html>`_
+of ``__array_ufunc__`` on ndarray subclasses (e.g., in astropy), but this
+would be acceptable given the still experimental status of``__array_ufunc__``.
```

Comment:
I might as well object loudly now, and be quite clear that it will need a much more convincing case than this to override my objections.